### PR TITLE
S/MIME: Canonicalize multi-part messages before hashing.  Improve error checking in wc_MIME_parse_headers.

### PR DESCRIPTION
**Pull request by kabuobeid**

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -54461,38 +54464,60 @@ WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in,
 
             section[0] = '\0';
             sectionLen = 0;
+            canonSection = XMALLOC((remainLen+1)*sizeof(char), NULL,
```
Comment (dgarske) [inline src/ssl.c:54467]: Sizeof char will always be 1.

File: wolfcrypt/src/asn.c [hunk 2]
```diff
@@ -18742,6 +18756,41 @@ MimeParam* wc_MIME_find_param_attr(const char* attribute,
     return param;
 }
 
+/*****************************************************************************
+* wc_MIME_canonicalize - Canonicalize a line by converting all line endings
```
Comment (dgarske) [inline wolfcrypt/src/asn.c:18760]: Is the reason for converting to CRLF is so the hash will always compute the same?

File: src/ssl.c [hunk 3]
```diff
@@ -54461,38 +54464,60 @@ WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in,
...
+            if (canonSection == NULL) {
+                goto error;
+            }
+
             lineLen = wolfSSL_BIO_gets(in, section, remainLen);
             while(XSTRNCMP(&section[sectionLen], boundary, boundLen) &&
                            remainLen > 0) {
+                canonLine = wc_MIME_canonicalize(&section[sectionLen]);
+                if (canonLine == NULL) {
+                    XFREE(canonSection, NULL, DYNAMIC_TYPE_PKCS7);
```
Comment (dgarske) [inline src/ssl.c:54478]: Maybe put this cleanup in the error label section?

File: src/ssl.c [hunk 4]
```diff
@@ -54461,38 +54464,60 @@ WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in,
...
+            ret = wolfSSL_BIO_write(*bcont, canonSection,
+                                    (int)XSTRLEN(canonSection));
+            if (ret != (int)XSTRLEN(canonSection)) {
+                XFREE(canonSection, NULL, DYNAMIC_TYPE_PKCS7);
                 goto error;
             }
             if ((bcontMemSz = wolfSSL_BIO_get_mem_data(*bcont, &bcontMem)) < 0) {
                 goto error;
             }
+            XFREE(canonSection, NULL, DYNAMIC_TYPE_PKCS7);
```
Comment (dgarske) [inline src/ssl.c:54520]: Set `canonSection` to NULL?

File: src/ssl.c (hunk 1 above)
Comment (dgarske) [inline src/ssl.c:54467]: The XMALLOC needs a cast to `(char*)` otherwise some compilers will complain.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/asn.c (hunk 2 above)
Comment (kabuobeid) [reply wolfcrypt/src/asn.c:18760]: In a multipart message, the first part is the content that is hashed (which we are converting here) and the second part is a PKCS7 object which includes a hash of the content.  The hash is calculated using the canonicalized (CRLF converted) content, not the exact content included.  So we need to convert it to calculate the hash correctly.

**Review by dgarske — APPROVED** [review]

File: certs/test/smime-test-canon.p7s [hunk 5]
```diff
@@ -0,0 +1,63 @@
+MIME-Version: 1.0
```
Comment (dgarske) [inline certs/test/smime-test-canon.p7s:1]: You may need to add `certs/test/smime-test-canon.p7s` to the include.am?

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]

**Pull request merged**